src/pygenstability/data_clustering.py

```diff
@@ -0,0 +1,286 @@
+        D_norm = D / np.amax(D)
+
+        # compute normalised similarity matrix
+        S = 1 - D_norm
+
+        # sparsify distance matrix with CkNN or kNN method
+        if self.method == "cknn":
+            sparse = compute_CkNN(D_norm, self.k, self.delta)
+
+        elif self.method == "knn":
+            sparse = compute_kNN(D_norm, self.k)
+
+        # undirected distance backbone is given by sparse graph and MST
+        mst = minimum_spanning_tree(D_norm)
+        backbone = np.array((mst + mst.T + sparse + sparse.T) > 0, dtype=int)
+
+        # apply distance threshold to backbone
+        backbone *= np.array(D <= self.distance_threshold, dtype=int)
+
+        # adjacency matrix has weights of similarity matrix
+        self.adjacency_ = S * backbone
+
+        return self.adjacency_
+
+
+class DataClustering(GraphConstruction):
+    """Data clustering."""
+
+    def __init__(
+        self,
+        metric="euclidean",
+        graph_method="cknn",
+        k=5,
+        delta=1.0,
+        distance_threshold=np.inf,
+        constructor="linearized",
+        min_scale=-3.0,
+        max_scale=0.0,
+        n_scale=50,
+        log_scale=True,
+        scales=None,
+        n_tries=100,
+        with_NVI=True,
+        n_NVI=20,
+        with_postprocessing=True,
+        with_ttprime=True,
+        with_spectral_gap=True,
+        exp_comp_mode="spectral",
+        result_file="results.pkl",
+        n_workers=4,
+        tqdm_disable=False,
+        with_optimal_scales=True,
+        optimal_scales_kwargs=None,
+        ms_method="louvain",
+        constructor_kwargs=None,
+    ):
+
+        # initialise parameters for graph construction
+        super().__init__(
+            metric=metric,
+            method=graph_method,
+            k=k,
+            delta=delta,
+            distance_threshold=distance_threshold,
+        )
+
+        # initialise parameters for PyGenStability
+        self.constructor = constructor
+        self.min_scale = min_scale
+        self.max_scale = max_scale
+        self.n_scale = n_scale
+        self.log_scale = log_scale
+        self.scales = scales
+        self.n_tries = n_tries
+        self.with_NVI = with_NVI
+        self.n_NVI = n_NVI
+        self.with_postprocessing = with_postprocessing
+        self.with_ttprime = with_ttprime
+        self.with_spectral_gap = with_spectral_gap
+        self.exp_comp_mode = exp_comp_mode
+        self.result_file = result_file
+        self.n_workers = n_workers
+        self.tqdm_disable = tqdm_disable
+        self.with_optimal_scales = with_optimal_scales
+        self.optimal_scales_kwargs = optimal_scales_kwargs
+        self.MS_method = ms_method
+        self.constructor_kwargs = constructor_kwargs
+
+        # attributes
+        self.adjacency_ = csr_matrix
+        self.results_ = {}
+        self.labels_ = []
+
+    def fit(self, X):
+
+        # construct graph
+        self.adjacency_ = csr_matrix(super().fit(X))
```

Comment:
would `self.fit(X)` not work?

Reply:
I thought we need the super() to call the fit method from the parent class, but not completely sure about this.

Reply:
you inherit from parent, so you have the parents method/attributes

Reply:
I think it might be necessary because we override a function that has the same name in the parent class. If I don't use the super() I get this error: `RecursionError: maximum recursion depth exceeded`

Reply:
see my new comments on renaming the fit() methods